packages/mds-cache/unflatteners.ts

```diff
@@ -1,11 +1,171 @@
+  UUID
+} from '@mds-core/mds-types'
 import {
   isStringifiedTelemetry,
   isStringifiedEventWithTelemetry,
   isStringifiedCacheReadDeviceResult
 } from '@mds-core/mds-schema-validators'
 
-import { StringifiedEvent, StringifiedTelemetry, StringifiedCacheReadDeviceResult, CachedItem } from './types'
+import log from '@mds-core/mds-logger'
+import { ParseError } from '@mds-core/mds-utils'
+import {
+  StringifiedEvent,
+  StringifiedTelemetry,
+  StringifiedCacheReadDeviceResult,
+  CachedItem,
+  StringifiedStateEntry,
+  StringifiedAllDeviceStates,
+  StringifiedTripsEvents,
+  StringifiedTripsTelemetry,
+  StringifiedAllTripsEvents
+} from './types'
+
+function parseDeviceState(deviceState: StringifiedStateEntry): StateEntry {
+  try {
+    return {
+      vehicle_type: deviceState.vehicle_type as VEHICLE_TYPE,
+      type: deviceState.type,
+      timestamp: Number(deviceState.timestamp) as Timestamp,
+      device_id: deviceState.device_id,
+      provider_id: deviceState.provider_id,
+      recorded: Number(deviceState.recorded) as Timestamp,
+      annotation_version: Number(deviceState.annotation_version),
+      annotation: deviceState.annotation
+        ? { in_bound: JSON.parse(deviceState.annotation.in_bound), areas: deviceState.annotation.areas }
+        : null,
+      gps: deviceState.gps
+        ? {
+            lat: Number(deviceState.gps.lat),
+            lng: Number(deviceState.gps.lng),
+            altitude: deviceState.gps.altitude ? Number(deviceState.gps.altitude) : null,
+            heading: deviceState.gps.heading ? Number(deviceState.gps.heading) : null,
+            speed: deviceState.gps.speed ? Number(deviceState.gps.speed) : null,
+            accuracy: deviceState.gps.accuracy ? Number(deviceState.gps.accuracy) : null
+          }
+        : null,
+      service_area_id: deviceState.service_area_id ? (deviceState.service_area_id as UUID) : null,
+      charge: deviceState.charge ? Number(deviceState.charge) : null,
+      state: deviceState.state ? (deviceState.state as VEHICLE_STATUS) : null,
+      event_type: deviceState.event_type ? (deviceState.event_type as VEHICLE_EVENT) : null,
+      event_type_reason: deviceState.event_type_reason ? (deviceState.event_type_reason as VEHICLE_REASON) : null,
+      trip_id: deviceState.trip_id ? (deviceState.trip_id as UUID) : null
+    }
+  } catch (err) {
+    throw new ParseError(`unable to parse deviceState: ${deviceState}`)
+  }
+}
+
+function parseAllDeviceStates(allDeviceStates: StringifiedAllDeviceStates): { [vehicle_id: string]: StateEntry } {
+  try {
+    const devices: { [vehicle_id: string]: StateEntry } = {}
+    Object.keys(allDeviceStates).map(vehicle_id => {
+      devices[vehicle_id] = parseDeviceState(allDeviceStates[vehicle_id])
+    })
+    return devices
+  } catch (err) {
+    throw new ParseError(`unable to parse allDeviceStates`)
+  }
+}
+
+async function parseTripsEvents(tripsEventsStr: StringifiedTripsEvents): Promise<TripsEvents> {
```

Comment:
If it's not too much trouble, could you add a couple quick unit tests to verify these transformations work? If possible, I'd also like to see some sample output showing all this stuff works.

Reply:
Also if the README could be updated with information on how to actually run this thing that'd be great.